feat: add DisableRateLimitCheck option to client

stevehipwell opened the pull request:
This PR adds the `DisableRateLimitCheck` option to the client to disable all client level rate limit logic. It also updates the documentation on the rate limit logic.

Fixes #3585

@gmlewis apologies if there are quality issues with this PR, my VS Code is broken for this codebase so I put off the work as long as I could.

gmlewis reviewed (approved):
Thank you, @stevehipwell!
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo.

alexandear commented on README.md:
```diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
-	log.Println("hit rate limit")
+var rateErr *github.RateLimitError
+if errors.As(err, &rateError) {
+  log.Printf("hit primary rate limit, used %d of %d\n", rateErr.Rate.Used, rateErr.rate.Limit)
```
```suggestion
	log.Printf("hit primary rate limit, used %d of %d\n", rateErr.Rate.Used, rateErr.Rate.Limit)
```

alexandear commented on README.md:
```diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
+up-to-date rate limit data for the client.
 
-To detect an API rate limit error, you can check if its type is `*github.RateLimitError`:
+To detect a primary API rate limit error, you can check if the error is a
+`RateLimitError`.
```
```suggestion
To detect a primary API rate limit error, you can check if the error is a `github.RateLimitError`.
```

stevehipwell replied:
This block was limited to 80 characters and I've kept it this way other than where there is a link (inspired by the Google Go formatting doc). Given that there is no linting or formatting rules and the README isn't consistent I don't think there is any value in being pedantic here (personally I prefer unbounded line lengths).

alexandear commented on README.md:
```diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
-This rate limit prevents clients from making too many concurrent requests.
-
-To detect an API secondary rate limit error, you can check if its type is `*github.AbuseRateLimitError`:
+To detect an API secondary rate limit error, you can check if the error is an
+`AbuseRateLimitError`.
```
```suggestion
To detect an API secondary rate limit error, you can check if the error is an `github.AbuseRateLimitError`.
```

stevehipwell replied:
Same as above.

alexandear commented on README.md:
```diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
-	log.Println("hit secondary rate limit")
+var rateErr *github.AbuseRateLimitError
+if errors.As(err, &rateError) {
+  log.Printf("hit secondary rate limit, retry after %v\n", rateErr.RetryAfter)
```
```suggestion
	log.Printf("hit secondary rate limit, retry after %v\n", rateErr.RetryAfter)
```

alexandear commented on README.md:
````diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
 ```
 
-Alternatively, you can block until the rate limit is reset by using the `context.WithValue` method:
+If you hit the primary rate limit, you can use the `SleepUntilPrimaryRateLimitReset`
````
I didn't find any `SleepUntilPrimaryRateLimitReset` method. Is it a mistake?

Maybe you mean `github.SleepUntilPrimaryRateLimitResetWhenRateLimited`?

stevehipwell replied:
> I didn't find any SleepUntilPrimaryRateLimitReset method. Is it a mistake?

It's not a method, the code below shows what it is and how to use it.

> Maybe you mean github.SleepUntilPrimaryRateLimitResetWhenRateLimited?

Everything in this module has a `github` prefix using the default import so I think this would be unnecessary.

alexandear replied:
What I mean, is `SleepUntilPrimaryRateLimitReset` != `SleepUntilPrimaryRateLimitResetWhenRateLimited`

alexandear commented on README.md:
```diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
-You can use [gofri/go-github-ratelimit](https://github.com/gofri/go-github-ratelimit) to handle
-secondary rate limit sleep-and-retry for you, as well as primary rate limit abuse-prevention and callback triggering.
+If you need to make a request even if the rate limit has been hit you can use
+the `BypassRateLimitCheck` method to bypass the rate limit check and make the
```
Let's add `github.` for consistency:
```suggestion
the `github.BypassRateLimitCheck` method to bypass the rate limit check and make the
```

alexandear commented on README.md:
```diff
@@ -189,54 +189,71 @@ using the installation ID of the GitHub app and authenticate with the OAuth meth
-["About secondary rate limits"](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits).
+If the client is an [OAuth app](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api#primary-rate-limit-for-oauth-apps)
+you can use the apps higher rate limit to request public data by using the
+`UnauthenticatedRateLimitedTransport` to make calls as the app instead of as
```
```suggestion
`github.UnauthenticatedRateLimitedTransport` to make calls as the app instead of as
```

alexandear commented on github/github.go:
```diff
@@ -171,6 +171,11 @@ type Client struct {
 	// User agent used when communicating with the GitHub API.
 	UserAgent string
 
+	// DisableRateLimitCheck stops the client checking for rate limits or tracking
+	// them. This is different to setting BypassRateLimitCheck in the context,
```
```suggestion
	// DisableRateLimitCheck stops the client checking for rate limits or tracking them.
	// This is different to setting BypassRateLimitCheck in the context,
```

alexandear commented on github/github.go:
```diff
@@ -850,21 +855,26 @@ func (c *Client) bareDo(ctx context.Context, caller *http.Client, req *http.Requ
-			return &Response{
-				Response: err.Response,
-			}, err
+	if !c.DisableRateLimitCheck {
```
Could we add a unit test for this new field?

stevehipwell commented:
@alexandear I've added a unit test for the new disable feature and one for the existing bypass feature that didn't have one.

@gmlewis I rebased to fix the merge conflict, I hope that's OK (I've left the commit structure as it was).

gmlewis commented:
Thank you, @stevehipwell and @alexandear!
Merging.
